text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+* An in-place, mutable `freeze` could be offered, e.g. `MaybeUninit::freeze(&mut self)`
+    * While this function would seem to be a simple body that llvm could replace with a runtime no-op, in reality it is possible for [virtual memory](https://man7.org/linux/man-pages/man2/madvise.2.html#MADV_FREE) that has been freshly allocated and has not been written to exhibit properties of uninitialized memory (rather than simply being an abstract phenomenon the compiler tracks that disappears at runtime). Thus, such an operation would require a potentially expensive in-place copy. Until such time as an optimized version is available, we should avoid defining the in-place version, and require users to spell it explicitly as `*self = core::mem::replace(&mut self, uninit()).freeze()`.
+    * If intermediate representations are cooperative, it may be beneficial to provide the operation in the future, as it could perform only the writes required to ensure the backing memory is in a stable state (such as 1 write every 4096 bytes)
+* `MaybeUninit<Int>` (and maybe `MaybeUninit<Float>`) could have arithmetic that is defined as `uninit (op) x` or `x (op) uninit` is `uninit`.
+    * While this can partially solve the second use case (by following it through `Simd`) and the third use case, this does not help the first use case
+    * This RFC does not preclude these operations from being defined in the future, and may even make them more useful.
+* `MaybeUninit::freeze` could instead be `pub unsafe fn freeze(self) -> T`, like `assume_init`.
+    * While this would allow `mu.freeze().assume_init()` to be written in fewer lines of code, maintaining the value as `MaybeUninit` may make some code possible using careful offsetting and `MaybeUninit::as_ptr()`/`MaybeUninit::as_mut_ptr()`.
+    * Most uses of `MaybeUninit` would require immediately using `.assume_init()` on the result, however, this is a potential footgun if `T` has any invalid initialized values, or is a user-defined type with a complex safety invariant. It is hoped that the extra verbosity, on top of helpful documentation, will allow intermediate, advanced, or expert unsafe code users making use of `MaybeUninit::freeze` to recognize this potential footgun and to carefully validate the types involved. 
+
+# Prior art
+[prior-art]: #prior-art
+
+[LLVM](https://llvm.org/docs/LangRef.html#freeze-instruction) supports freeze by-value. 
+
+
+* GCC may support an equivalent operation via the `SAVE_EXPR` GIMPLE code.
+* See <https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/GCC.20and.20freeze>
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* Which of the library functions should recieve the direct language intrinsic, between `ptr::read_freeze` and `MaybeUninit::freeze`
```

Comment:
A third option is to simply stabilize the intrinsic in-situ, like [`core::intrinsics::copy`](https://doc.rust-lang.org/std/intrinsics/fn.copy.html).

Reply:
That was accidental (only `core::ptr::copy` was supposed to be stable but one is a re-export of the other and rustc used to not be able to distinguish re-exports for stability). I don't think we should repeat it.

Reply:
Agreed. I don't want any more directly exposed intrinsics.